### PR TITLE
✨ Add dirtyElement method to Resources

### LINKED ISSUE #14686: Layers: Expose element dirty method
Certain events require DOM manipulations in the same process tick to work (eg, https://github.com/ampproject/amphtml/pull/14676).

So, we can't wait for a vsync mutate phase. But, if we mutate immediately, the cache will still provide invalid values until it's busted in the next mutate.

So, I need to add a method to dirty the cache immediately for these special cases.

### PR DESCRIPTION
**Pull request by jridgewell**
Fixes #14686.

/cc @cathyxz

### CODE REVIEW

File: src/service/resources-impl.js [hunk 1]
```diff
@@ -1012,19 +1012,40 @@ export class Resources {
...
-        this.layers_.dirty(element);
+        this.dirtyElement(element);
       },
     });
   }
 
+  /**
+   * Dirties the cached element measurements after a mutation occurs.
+   *
+   * TODO(jridgewell): This API needs to be audited. Common practice is
```
Comment (cathyxz) [inline src/service/resources-impl.js]: Is there currently a way to tell layers that we're only mutating children of an element and not necessarily the entire element itself?

Comment (jridgewell) [reply src/service/resources-impl.js]: You pass in the element being mutated.

```js

this.mutateElement(() => {
  // do something to child.
}, child);
```

File: src/service/resources-impl.js [hunk 2]
```diff
@@ -1012,19 +1012,40 @@ export class Resources {
...
+  dirtyElement(element) {
+    if (this.useLayers_) {
+      this.layers_.dirty(element);
+    } else {
+      const isAmpElement = element.classList.contains('i-amphtml-element');
+      if (isAmpElement) {
+        const r = Resource.forElement(element);
+        this.setRelayoutTop_(r.getLayoutBox().top);
+      }
+      this.schedulePass(FOUR_FRAME_DELAY_, !isAmpElement)
```
Comment (cathyxz) [inline src/service/resources-impl.js]: semicolon =)

File: src/service/resources-impl.js [hunk 3]
```diff
@@ -1012,19 +1012,40 @@ export class Resources {
     return this.vsync_.runPromise({
       measure: measurer || undefined,
       mutate: () => {
+        // TODO: Audit this system. Did this cause a layer invalidation (new
```
Comment (choumx) [inline src/service/resources-impl.js]: `TODO(jridgewell)`

File: src/service/resources-impl.js [hunk 4]
```diff
@@ -1012,19 +1012,40 @@ export class Resources {
...
+  dirtyElement(element) {
+    if (this.useLayers_) {
+      this.layers_.dirty(element);
+    } else {
+      const isAmpElement = element.classList.contains('i-amphtml-element');
+      if (isAmpElement) {
+        const r = Resource.forElement(element);
+        this.setRelayoutTop_(r.getLayoutBox().top);
+      }
+      this.schedulePass(FOUR_FRAME_DELAY_, !isAmpElement);
```
Comment (choumx) [inline src/service/resources-impl.js]: This change in behavior is fixing a non-layers bug? I don't see it in the referenced issue.

Comment (cathyxz) [reply src/service/resources-impl.js]: It's due to https://github.com/ampproject/amphtml/pull/14676. Basically there are situations where we cannot put a mutation in an async callback (i.e. mutateElement). Therefore we need to mutate it and then call `dirtyElement` to tell layers that the element was mutated.

**Comment by jridgewell** [comment]
Need approval.

**Pull request merged**